lib/lcevc/lcevc_dil.js

```diff
@@ -0,0 +1,277 @@
+/*! @license
+ * Shaka Player
+ * Copyright 2016 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+goog.provide('shaka.lcevc.Dil');
+goog.require('shaka.log');
+goog.require('shaka.util.Dom');
+goog.require('shaka.util.IReleasable');
+
+/**
+ * @summary
+ *  lcevcDil - (MPEG-5 Part 2 LCEVC - Decoder Integration Layer) provides
+ *  all the operations related to the enhancement and rendering
+ *  of LCEVC enabled streams and on to a canvas.
+ * @implements {shaka.util.IReleasable}
+ * @export
+ */
+shaka.lcevc.Dil = class {
+  /**
+   * @param {HTMLVideoElement} media The video element that will be attached to
+   * LCEVC Dil for input.
+   * @param {HTMLCanvasElement} canvas The canvas element that will be attached
+   * to LCEVC Dil to render the enhanced frames.
+   * @param {LcevcDil.LcevcDilConfig=} dilConfig The LCEVC DIL
+   * config object to initialize the LCEVC DIL.
+   */
+  constructor(media, canvas, dilConfig) {
+    /** @private {?LcevcDil.LcevcDIL} */
+    this.dil_ = null;
+
+    /** @private {?Element} */
+    this.parent_ = null;
+
+    /** @private {?HTMLElement} */
+    this.lcevcContainer_ = null;
+
+    /** @private {ResizeObserver} */
+    this.resizeObserver_ = null;
+
+    this.level_ = -1;
+    this.containerFormat_ = 0;
+
+    /** @type {HTMLVideoElement} */
+    this.media_ = media;
+
+    /** @type {HTMLCanvasElement} */
+    this.canvas_ = canvas;
+
+    /** @type {LcevcDil.LcevcDilConfig} */
+    this.dilConfig_ = dilConfig ? dilConfig :
+        shaka.lcevc.Dil.DefaultLcevcDilConfig;
+
+    this.createDil_();
+  }
+
+  /**
+   * Append data to the LCEVC Dil.
+   * @param {BufferSource} data
+   */
+  appendBuffer(data) {
+    if (this.dil_) {
+      this.dil_.appendBuffer(data, 'video', this.level_);
+    }
+  }
+
+  /**
+  * Hide the canvas and show original video element
+  */
+  hideCanvas() {
+    if (this.dil_) {
+      this.media_.style.display = 'block';
+      this.canvas_.style.display = 'none';
+    }
+  }
+
+  /**
+    * Show the canvas and hide original video element
+    */
+  showCanvas() {
+    if (this.dil_) {
+      this.media_.style.display = 'none';
+      this.canvas_.style.display = 'block';
+    }
+  }
+
+  /**
+   * Create LCEVC Dil.
+   * @private
+   */
+  createDil_() {
+    if (this.isSupported_() && !this.dil_) {
+      if (!this.canvas_) {
+        this.createCanvas_();
```

Comment:
I really don't want you to create a canvas at this level.  When the Shaka UI isn't in use, applications should always provide their own DOM elements (through the Player constructor or a method on Player).  We won't create any, because it gets messy quickly if apps aren't in control of their own layout.

At the UI layer, though, we will create and style elements automatically.  At that level, we should create a canvas by default and provide it to the Player constructor.  See ui/ui.js, specifically setupUIandAutoLoad_.  You can create a canvas, append it to the container, and give it a class name like "shaka-canvas", with whatever default styles you want in ui/less/containers.less.

Reply:
Created a new canvas in setupUIandAutoLoad_ which is used if there is no canvas provided from the user in the constructor. Changed the function to `getCanvas_()` so that lib can use the canvas from UI can be used. Please let me know if that is ok.